docs/proposal/20220929-ipam-support.md

```diff
@@ -0,0 +1,423 @@
+      - [VSphereVM Controller Changes](#vspherevm-controller-changes)
+      - [Owner Reference](#owner-reference)
+      - [Flavors](#flavors)
+      - [Validation](#validation)
+      - [Nameservers](#nameservers)
+    - [Security Model](#security-model)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+    - [Version Skew Strategy](#version-skew-strategy)
+  - [References](#references)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API IPAM Integration Glossary](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220125-ipam-integration.md#glossary)
+
+## Summary
+
+This proposal adds CAPV support for the IPAM API contract described in the
+[IPAM Integration
+proposal](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220125-ipam-integration.md).
+CAPV will be able to allocate/release IP Addresses from an IPAM provider's IP
+Pool to be used by VSphereMachines.
+
+## Motivation
+
+The initial IPAM integration proposal explains the motivation for IPAM pretty well:
+
+> IP address management for machines is currently left to the infrastructure
+> providers for CAPI. Most on-premise providers (e.g. CAPV) allow the use of
+> either DHCP or statically configured IPs. Since Machines are created from
+> templates, static allocation would require a single template for each machine,
+> which prevents dynamic scaling of nodes without custom controllers that create
+> new templates.
+>
+> While DHCP is a viable solution for dynamic IP assignment which also enables
+> scaling, it can cause problems in conjunction with CAPI. Especially in smaller
+> networks rolling cluster upgrades can exhaust the network in use. When multiple
+> machines are replaced in quick succession, each of them will get a new DHCP
+> lease. Unless the lease time is very short, at least twice as many IPs as the
+> maximum number of nodes has to be allocated to a cluster.
+>
+> Metal3 has an ip-address-manager component that allows for in-cluster
+> management of IP addresses through custom resources, allowing to avoid DHCP.
+> CAPV allows to ommit the address from the template while having DHCP disabled,
+> and will wait until it is set manually, allowing to implement custom
+> controllers to take care of IP assignments. At DTAG we've extended metal3's
+> ip-address-manager and wrote a custom controller for CAPV to integrate both
+> with Infoblox, our IPAM solution of choice. The CAPV project has shown
+> interest, and there has been a proposal to allow integrating metal3's
+> ip-address-manager with external systems.
+>
+> All on-premise providers have a need for IP Address Management, since they
+> can't leverage SDN features that are available in clouds such as AWS, Azure or

```

Comment:
Hi! Why can't we leverage SDN features of cloud providers?

Reply:
This is a quote from the CAPI IPAM proposal, not written specifically for this one.

When you are on-premise, you can't use a cloud provider's SDN. It might be a viable use case to run vSphere on a cloud, and integrate it with the provider's SDN, but that's not the case for on-prem setups. The features I'm talking about here is the ability for the compute service of the cloud to request addresses from the networking service, and assign them statically. Even if you have SDN in your datacenter, you still need to integrate with it, which is something the IPAM contract allows you to do.